Inferred service dependencies guide changes (#23219)
DataDog/documentation: content/en/tracing/guide/inferred-service-opt-in.md

---

{{< callout url="https://docs.google.com/forms/d/1imGm-4SfOPjwAr6fwgMgQe88mp4Y-n_zV0K3DcNW4UA/edit" d_target="#signupModal" btn_hidden="true" btn_hidden="false" header="Opt in to the private beta!" >}}
Inferred service dependencies are in private beta. To request access, complete the form.
{{< /callout >}}

> ## Inferred service dependencies

Reviewer: ```suggestion
## Overview
```

> Datadog can automatically discover the dependencies for an instrumented service, such as a database, a queue, or a third-party API even if that dependency hasn't been instrumented yet. Datadog infers the presence of these dependencies and collects its associated performance metrics based on information on the outbound requests from your instrumented services.

Reviewer: ```suggestion
Datadog can automatically discover the dependencies for an instrumented service, such as a database, a queue, or a third-party API, even if that dependency hasn't been instrumented yet. By analyzing outbound requests from your instrumented services, Datadog infers the presence of these dependencies and collects associated performance metrics.
```
Minor suggestions for clarity.

> With the new inferred entities experience, filter [Service Catalog][3] entries by entity type (database, queue, third-party API, etc), and better visualize service dependencies from the new [Service Page dependency map](#service-page-dependency-map) and other places in the APM product experience.

Reviewer: ```suggestion
With the new inferred entities experience, you can filter [Service Catalog][3] entries by entity type, such as database, queue, or third-party API. This allows you to better visualize service dependencies using the [Service Page dependency map](https://github.com/DataDog/documentation/pull/23219/files#service-page-dependency-map) and APM features.
```

> To determine the names and types of the inferred service dependencies, Datadog uses standard span attributes and maps them to `peer.*` attributes - find the full list of `peer.*` attributes [here](#inferred-service-dependencies-nomemclature) -. Inferred external APIs use the default naming scheme `net.peer.name`. For example, `api.stripe.com`, `api.twilio.com`, `us6.api.mailchimp.com`. Inferred databases use the default naming scheme `db.instance`.

Reviewer: ```suggestion
To determine the names and types of the inferred service dependencies, Datadog uses standard span attributes and maps them to `peer.*` attributes. For the full list of `peer.*` attributes, see [Inferred service dependencies nomenclature](#inferred-service-dependencies-nomemclature). Inferred external APIs use the default naming scheme `net.peer.name`. For example, `api.stripe.com`, `api.twilio.com`, `us6.api.mailchimp.com`. Inferred databases use the default naming scheme `db.instance`.
```

If you're using the Go, Java, NodeJS, PHP, .NET, or Ruby tracer, you can customize the default names for inferred entities. 

**Note:** If you configure monitors, dashboards, or notebooks for a given inferred service during the beta, you may need to update them if the naming scheme changes. Read more about migration steps in the [opt-in instructions](#opt-in).

### Service page Dependency map

Use the dependency map to visualize service-to-service communication and gain insight into system components such as databases, queues, and third-party dependencies. You can group dependencies by type and filter by Requests, Latency, or Errors to identify slow or failing connections.

{{< img src="tracing/services/service_page/dependencies.png" alt="Service page service dependency map" style="width:100%;">}}

## Opt in

<div class="alert alert-warning">Only go through migration steps once Datadog support confirmed the feature is enabled for you on the Datadog side.</div>

To opt in, you must adjust:
- your [Datadog Agent](#datadog-agent-configuration) (or [OpenTelemetry collector](#opentelemetry-collector)) configuration

> - your [APM tracing libraries](#apm-tracing-libary-configuration) configuration

Reviewer 2: This is optional but recommended. If a customer doesn't want to take on the breaking changes, the experience still works but not as well.

> To opt in, you must adjust:
> - your [Datadog Agent](#datadog-agent-configuration) (or [OpenTelemetry collector](#opentelemetry-collector)) configuration
> - your [APM tracing libraries](#apm-tracing-libary-configuration) configuration

Reviewer: ```suggestion
To opt in, Datadog recommends you adjust your:
- [Datadog Agent](#datadog-agent-configuration) (or [OpenTelemetry collector](#opentelemetry-collector)) configuration
- [APM tracing libraries](#apm-tracing-libary-configuration) configuration
```
Simplify repetition in bullets and indicate this isn't required per Peter's note.

{{< /code-block >}}

### APM tracing libary configuration

> <div class="alert alert-warning">Going through the following step is a <b>breaking change</b>: Datadog will change the way service names are captured by default. Check the <a href="https://docs.datadoghq.com/tracing/guide/inferred-service-opt-in#global-default-service-naming-migration">Global default service naming migration</a>, to see if you need to take any migration actions.</div>

Reviewer: ```suggestion
<div class="alert alert-warning">The following steps introduce a <b>breaking change</b>: Datadog will change the way service names are captured by default. Refer to <a href="#global-default-service-naming-migration">Global default service naming migration</a>, to determine if you need to take any migration actions.</div>
```

{{< /tabs >}}

> ## The new nomemclature: what is changing

Reviewer: ```suggestion
## The new nomenclature: What is changing
```

### List of newly introduced peer.* tags 

### Global default service naming migration

When you enable the `DD_TRACE_REMOVE_INTEGRATION_SERVICE_NAMES_ENABLED` environment variable, it improves how service-to-service connections and inferred services are represented in Datadog visualizations, across all supported tracing library languages and integrations.

Previously, some tracing libraries included the name of the associated integration in service name tagging. For example, .NET tagged gRPC calls as `service:<DD_SERVICE>-grpc-client` while Python tagged them as `service:grpc-client`. With this option enabled, all supported tracing libraries tag spans from the downstream services with the calling service's name, `service:<DD_SERVICE>`, thereby providing a _global default service name_.

_ | Before | After
--|-------|--------
Service name | `service:my-service-grpc-client` or `service:grpc-client` | `service:myservice` 

> additional `peer.*` attributes | _No `peer.*` tags set_ | `@peer.service:otherservice` (`otherservice` being the name of the remote service being called via gRPC)

Reviewer: ```suggestion
additional `peer.*` attributes | _No `peer.*` tags set_ | `@peer.service:otherservice` (`otherservice` being the name of the remote service being called with gRPC)
```

Similarly, for a span representing a call to a mySQL database:

_ | Before | After
--|-------|--------
Service name | `service:my-service-mysql` or `service:mysql` | `service:myservice` 
additional `peer.*` attributes | _No `peer.*` tags set_ | `@peer.db.name:user-db`, `@peer.db.system:mysql`

Consequently, if you have existing:

- APM metrics
- APM custom span metrics
- Trace analytics
- Retention filters
- Sensitive data scans
- Monitors, dashboards, or notebooks 

> ... that target similar service names, update those items to use the global default service tag (`service:<DD_SERVICE>`) instead.

Reviewer: ```suggestion
And these target similar service names, update those items to use the global default service tag (`service:<DD_SERVICE>`) instead.
```
